File: Hot 100/131. 分割回文串.cpp
class Solution {
public:
    bool isPalindrome(string &s, int start, int end) {
        while(start < end) {
            if(s[start] != s[end]) {
                return false;
            }
            start++;
            end--;
        }
        return true;
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string> > ans;
        vector<string> path;
        int n=s.length();
        function<void(int)> dfs = [&](int i) {
            if(i==n) {
                ans.push_back(path);
                return;
            }
            for(int j=i; j<n; j++) {
                if(isPalindrome(s, i ,j)) {
                    path.push_back(s.substr(i, j-i+1));
                    dfs(j+1);
                    path.pop_back();
                }
            }
        };
        dfs(0);
        return ans;
    }
};

File: 刷题/Hot 100/131. 分割回文串.go
func isPlindrome(s string, start int, end int) bool {
    for start < end {
        if s[start] != s[end] {
            return false
        }
        start++
        end--
    }
    return true
}

func partition(s string) [][]string {
    ans:=[][]string{}
    path:=[]string{}
    n:=len(s)
    var dfs func(int)
    dfs = func(i int) {
        if i == n {
            ans = append(ans, append([]string(nil), path...))
            return
        }
        for j:=i; j<n; j++ {
            if isPlindrome(s, i, j) {
                path = append(path, s[i:j+1])
                dfs(j+1)
                path = path[:len(path)-1]
            }
        }
    }
    dfs(0)
    return ans
}